primitives/trie/src/storage_proof.rs

```diff
@@ -14,65 +14,990 @@
+/// a storage proof.
+/// Some kind are reserved for test or internal use and will
+/// not be usable when decoding proof.
+#[repr(u8)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum StorageProofKind {
+	/// Kind for `StorageProof::Flatten`.
+	Flatten = 1,
+
+	/// Kind for `StorageProof::TrieSkipHashes`.
+	TrieSkipHashes = 2,
+
+	/// Kind for `StorageProof::KnownQueryPlanAndValues`.
+	KnownQueryPlanAndValues = 3,
+
+	/// Technical only
+
+	/// Kind for `StorageProof::TrieSkipHashesForMerge`.
+	TrieSkipHashesForMerge = 125,
+
+	/// Testing only indices
+
+	/// Kind for `StorageProof::Full`.
+	Full = 126,
+
+	/// Kind for `StorageProof::TrieSkipHashesFull`.
+	TrieSkipHashesFull = 127,
+}
+
+impl StorageProofKind {
+	/// Decode a byte value representing the storage kind.
+	/// Return `None` if the kind does not exists or is not allowed.
+	#[cfg(test)]
+	pub fn read_from_byte(encoded: u8) -> Option<Self> {
+		Some(match encoded {
+			x if x == StorageProofKind::Flatten as u8 => StorageProofKind::Flatten,
+			x if x == StorageProofKind::TrieSkipHashes as u8 => StorageProofKind::TrieSkipHashes,
+			x if x == StorageProofKind::KnownQueryPlanAndValues as u8
+				=> StorageProofKind::KnownQueryPlanAndValues,
+			x if x == StorageProofKind::Full as u8 => StorageProofKind::Full,
+			x if x == StorageProofKind::TrieSkipHashesFull as u8 => StorageProofKind::TrieSkipHashesFull,
+			x if x == StorageProofKind::TrieSkipHashesForMerge as u8
+				=> StorageProofKind::TrieSkipHashesForMerge,
+			x if x == StorageProofKind::TrieSkipHashesFull as u8
+				=> StorageProofKind::TrieSkipHashesFull,
+			_ => return None,
+		})
+	}
+
+	/// Decode a byte value representing the storage kind.
+	/// Return `None` if the kind does not exists or is not allowed.
+	#[cfg(not(test))]
+	pub fn read_from_byte(encoded: u8) -> Option<Self> {
+		Some(match encoded {
+			x if x == StorageProofKind::Flatten as u8 => StorageProofKind::Flatten,
+			x if x == StorageProofKind::TrieSkipHashes as u8 => StorageProofKind::TrieSkipHashes,
+			x if x == StorageProofKind::KnownQueryPlanAndValues as u8
+				=> StorageProofKind::KnownQueryPlanAndValues,
+			_ => return None,
+		})
+	}
+}
+
+#[derive(Clone)]
+/// Additional information needed for packing or unpacking storage proof.
+/// These do not need to be part of the proof but are required
+/// when processing the proof.
+pub enum Input {
+	/// Proof is self contained.
+	None,
+
+	/// Contains trie roots used during proof processing.
+	ChildTrieRoots(ChildrenProofMap<Vec<u8>>),
+
+	/// Contains trie roots used during proof processing.
+	/// Contains key and values queried during the proof processing.
+	QueryPlanWithValues(ChildrenProofMap<(Vec<u8>, Vec<(Vec<u8>, Option<Vec<u8>>)>)>),
+
+	/// Contains trie roots used during proof processing.
+	/// Contains keys queried during the proof processing.
+	QueryPlan(ChildrenProofMap<(Vec<u8>, Vec<Vec<u8>>)>),
+}
+
+impl Input {
+	/// Update input with new content.
+	/// Return false on failure.
```

Comment:
Why not return a `Result` instead? That's a clear pattern that has multiple syntactic benefits.